modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbIndexes.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.ignite.internal.storage.index.SortedIndexStorage;
+import org.apache.ignite.internal.storage.index.StorageHashIndexDescriptor;
+import org.apache.ignite.internal.storage.index.StorageIndexDescriptorSupplier;
+import org.apache.ignite.internal.storage.index.StorageSortedIndexDescriptor;
+import org.apache.ignite.internal.storage.rocksdb.index.AbstractRocksDbIndexStorage;
+import org.apache.ignite.internal.storage.rocksdb.instance.SharedRocksDbInstance;
+import org.jetbrains.annotations.Nullable;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.WriteBatch;
+
+/** Manager for all RocksDB-based indexes. */
+class RocksDbIndexes {
+    /** Hash Index storages by Index IDs. */
+    private final ConcurrentMap<Integer, HashIndex> hashIndices = new ConcurrentHashMap<>();
+
+    /** Sorted Index storages by Index IDs. */
+    private final ConcurrentMap<Integer, SortedIndex> sortedIndices = new ConcurrentHashMap<>();
+
+    private final SharedRocksDbInstance rocksDb;
+
+    /** Callback for getting partition storages using partition IDs. */
+    private final IntFunction<RocksDbMvPartitionStorage> partitionStorageProvider;
+
+    RocksDbIndexes(SharedRocksDbInstance rocksDb, IntFunction<RocksDbMvPartitionStorage> partitionStorageProvider) {
+        this.rocksDb = rocksDb;
+        this.partitionStorageProvider = partitionStorageProvider;
+    }
+
+    void recoverIndexes(StorageIndexDescriptorSupplier indexDescriptorSupplier) throws RocksDBException {
+        try (WriteBatch writeBatch = new WriteBatch()) {
+            for (int indexId : rocksDb.hashIndexIds()) {
+                var descriptor = (StorageHashIndexDescriptor) indexDescriptorSupplier.get(indexId);
+
+                if (descriptor == null) {
+                    deleteByPrefix(writeBatch, rocksDb.hashIndexCf(), createKey(BYTE_EMPTY_ARRAY, indexId));
+                } else {
+                    hashIndices.put(indexId, new HashIndex(rocksDb, descriptor, rocksDb.meta));
+                }
+            }
+
+            var indexCfsToDestroy = new ArrayList<Map.Entry<Integer, ColumnFamily>>();
+
+            for (Map.Entry<Integer, ColumnFamily> e : rocksDb.sortedIndexes().entrySet()) {
+                int indexId = e.getKey();
+
+                ColumnFamily indexCf = e.getValue();
+
+                var descriptor = (StorageSortedIndexDescriptor) indexDescriptorSupplier.get(indexId);
+
+                if (descriptor == null) {
+                    deleteByPrefix(writeBatch, indexCf, createKey(BYTE_EMPTY_ARRAY, indexId));
+
+                    indexCfsToDestroy.add(e);
+                } else {
+                    sortedIndices.put(indexId, SortedIndex.restoreExisting(rocksDb, indexCf, descriptor, rocksDb.meta));
+                }
+            }
+
+            rocksDb.db.write(DFLT_WRITE_OPTS, writeBatch);
+
+            if (!indexCfsToDestroy.isEmpty()) {
+                rocksDb.flusher.awaitFlush(false)
```

Comment:
Maybe we need to do a flush after we destroy the indexes too or not?

Reply:
I don't think we need to force it